Extract common methods from AMP and HTML exporters

nicolas-fricke opened the pull request:
AMP is basically just a custom version of HTML. Therefore, the code of the two exporters is very similar. For better reusability (e.g. when implementing a Facebook Instant Article exporter), extract the code into a `Common` module which then is reused by both classes.

For reviewing, the code in the `Common` modules is extracted from the `HTML` exporters. It should be fairly straight forward to review this commit-by-commit as each commit extracts the common code into a module and reuses it.

The only changes that were made to the base code other than extracting the methods are:
- Use `.exporter_by_type` instead of referencing elements directly (this is also better since this way the elements can be overwritten via the configuration)
- Add a `.base_class?` check to see if an element is the instance of the base class since comparing the class is not working anymore once it is extracted to a module
- Add `Base#base_class` to get the base class of an element instance
- Add `.export_format` to check for custom element exporters so that the lookup method can be made reusable

nicolas-fricke committed:
Extract common code from HTML and AMP `Base` exporters

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

For this, also:
- Add a `namespace` methods to return the module nesting of the base
  class and its children
- Add a `base_class?` check to see if an element is the instance of the
  base class since comparing the class is not working anymore once it
  is extracted to a module
- Add `export_format` to check for custom element exporters so that
  the lookup method can be made reusable

nicolas-fricke committed:
Extract common code from HTML and AMP `Text` exporters

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke committed:
Extract common code from HTML and AMP `Paragraph` exporters

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke commented:
[B]locked – This is a proposal for now and should be the basis of discussion. If we decide that extracting code like this is useful, I would do the same for all the other element classes.

towanda commented on lib/article_json/export/common/html/elements/base.rb:
```diff
@@ -0,0 +1,85 @@
+            # namespace, a simple `self.class == Base` check does not work.
+            # @return [Boolean]
+            def is_base_class?
+              self.class.name.end_with?('::Base')
```
Nicer than I thought!

nicolas-fricke committed:
Extract common code from HTML and AMP `Heading` exporter

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke committed:
Extract common code from HTML and AMP `List` exporter

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke committed:
Extract common code from HTML and AMP shared modules

AMP is basically just a custom version of HTML. Therefore, the code for
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both exporters.

towanda commented:
+1, I like it!

dsager commented on lib/article_json/export/common/html/elements/base.rb:
```diff
@@ -0,0 +1,85 @@
+            module ClassMethods
+              # Instantiate the correct sub class for a given element
+              # @param [ArticleJSON::Elements::Base] element
+              # @return [ArticleJSON::Export::Common::HTML::Elements::Base]
```
Is this the right way to say "it returns an object implementing this module"?

nicolas-fricke replied:
I think so – any other ideas?

dsager replied:
nope :)

towanda replied:
for me sounds good

dsager commented on lib/article_json/export/common/html/elements/base.rb:
```diff
@@ -0,0 +1,85 @@
+              # Look up the correct exporter class based on the element type
+              # @param [Symbol] type
+              # @return [ArticleJSON::Export::Common::HTML::Elements::Base]
+              def exporter_by_type(type)
```
The line breaks in this method feel weird. what about this?
```ruby
def exporter_by_type(type)
  key = type.to_sym
  custom_exporter = 
    ArticleJSON.configuration.element_exporter_for(export_format, key)
  custom_exporter || default_exporter_mapping[key]
end
```

nicolas-fricke replied:
This also doesn't fit within the 80 character limit due to the 8 level deep indentation of this method. But found another way to split it up now – please check if you like it better now :)

dsager replied:
yes, better! thanks!

dsager reviewed (approved):
Nice, definitely looks better than the duplication.

nicolas-fricke committed:
Extract common code from HTML and AMP `Quote` exporter

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

For this, add a `base_class` method to the shared `Base` module. This
method returns the base class for a given element.

nicolas-fricke committed:
Extract common code from HTML and AMP `TextBox` exporter

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke committed:
Extract common code from HTML and AMP `Image` exporter

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke committed:
Extract common code from HTML and AMP `Embed` exporter

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

nicolas-fricke commented:
@towanda @dsager Thanks for your positive feedback up to now :)

Based on this, I have now extracted the code of all AMP / HTML elements. Please have a look at it again and tell me if you approve of the changes. As mentioned in the description above, it's probably easiest to review this on a commit basis.

nicolas-fricke committed:
Extract common code from HTML and AMP exporters

AMP is basically just a custom version of HTML. Therefore, the code of
the two exporters is very similar. For better reusability (e.g. when
implementing a Facebook Instant Article exporter), extract the code into
a `Common` module which then is reused by both classes.

For this, also add a `.namespace` method to the extractor classes.

christofdamian commented:
+1 better than the duplication, though I am not a big fan of includes.

dsager commented on lib/article_json/export/common/html/elements/shared/caption.rb:
```diff
@@ -0,0 +1,31 @@
+              private
+
+              # @return [ArticleJSON::Export::Common::HTML::Elements::Base]
+              def text_class
```
The name is a bit too generic for my taste, given that it's from a module. A class that uses this module could have a method `text_class` referring to CSS classes. So maybe something like `caption_text_exporter` or `caption_text_exporter_class`?

nicolas-fricke replied:
Good point, I'll change this right away :)

nicolas-fricke replied:
Done :)

dsager commented on lib/article_json/export/common/html/shared/module_utils.rb:
```diff
@@ -0,0 +1,24 @@
+    module Common
+      module HTML
+        module Shared
+          module ModuleUtils
```
I don't like having this in a separate module much. Took me quite a bit to understand what's going on and what the name `ModuleUtils` actually means :)

I would just leave it in the exporter module where it s used, much easier to grasp than a module loading another module that extends the base class :)

nicolas-fricke replied:
The issue is that this is used in two modules: In the `Exporter` as well as in the `Elements::Base`. I first had the method duplicated in both modules, but that also didn't look right. Any better implementation ideas or naming suggestions?

dsager replied:
Like discussed in person, better to define explicit `namespace` methods where needed...

dsager reviewed (approved):
GJ! Just two comments on the code...

The one thing I don't like about this approach, now that I see it, is: The HTML export is basically empty and all its logic is in the common modules. Feels weird. But still better than the duplication and I can't think of a better way...

nicolas-fricke commented:
@dsager I tried to incorporate your ideas – please have another look at the PR 😃

dsager reviewed (approved):
Looks good. A bit tricky to review because of the rebase. In the future it's maybe better to add separate commits and do a rebase/fixup right before merging the PR, especially if it's such a big one like this...

towanda commented:
+1, much better than duplication, GJ!

nicolas-fricke commented:
@dsager Good point, I'll work with `fixup` commits next time :)

Thanks for the reviews!
